### DISCUSSION TITLE
How to filter by tenant Serenity.Net 5

### DISCUSSION THREAD

**Discussion by jovdej** [serenity-is/Serenity#5425] [Help]
How to filter by tenant (example Tutorials Multi Tenancy) because Serenity.Authorization static
class is removed from Serenity.Net 5.

private class MyRetrieveHandler: RetrieveRequestHandler
{
protected override void PrepareQuery (SqlQuery query)
{
base.PrepareQuery (query);

     var user = (UserDefinition) Authorization.UserDefinition;
     if (! Authorization.HasPermission (PermissionKeys.Tenants))
         query.Where (fld.TenantId == user.TenantId);
 }
}
https://serenity.is/docs/tutorials/multi_tenancy/preventing_edits_to_users_from_other_tenants

**Comment by erehtolleh** [comment]
Look at my reply for #5504 .  If you click on view more, it shows you how I injected IUserRetrieveService to get the UserDefinition

**Reply by minhhungit** [reply]
I use thid code

```csharp
var username = UserAccessor.User?.Identity?.Name;

if (UserRetriever.ByUsername(username) is UserDefinition user)
{
    if (!Permissions.HasPermission(Administration.PermissionKeys.Tenants))
        query.Where (fld.TenantId == user.TenantId);
}          
```

You will need to inject `IUserAccessor`, `IUserRetrieveService` and even `IPermissionService` first, for example:
```csharp
protected IUserAccessor UserAccessor { get; }
protected IUserRetrieveService UserRetriever { get; }
protected IPermissionService Permissions { get; }

public YourClass(
    IUserAccessor userAccessor, 
    IUserRetrieveService userRetriever,
    IPermissionService permissions)

{
    UserAccessor = userAccessor ?? throw new ArgumentNullException(nameof(userAccessor));
    UserRetriever = userRetriever ?? throw new ArgumentNullException(nameof(userRetriever));
    Permissions = permissions ?? throw new ArgumentNullException(nameof(permissions));
}
```

**Reply by imhemayatsangin** [reply]
dear @minhhungit  where exactly we can inject IUserAccessor, IUserRetrieveService and even IPermissionService in which file?

**Reply by minhhungit** [reply]
@sangincode404 your class, which you are trying to use above code. For this post, it is repository class 🙂

**Reply by imhemayatsangin** [reply]
Thanks @minhhungit it's working now 🙂
